File: packages/agent-python/shieldrasp/detection/ssrf.py
import socket
from urllib.parse import urlparse
from shieldrasp.taint.context import is_tainted

BLOCKED_RANGES = ["127.", "10.", "172.16.", "192.168.", "169.254."]

def detect_ssrf(url: str):
    if not is_tainted(url):
        return {"blocked": False, "matched": False}
        
    parsed = urlparse(url)
    hostname = parsed.hostname
    
    if not hostname:
        return {"blocked": False, "matched": False}
        
    if hostname == "localhost" or hostname.endswith(".internal"):
        return _block(url)
        
    try:
        ip = socket.gethostbyname(hostname)
        for r in BLOCKED_RANGES:
            if ip.startswith(r):
                return _block(url)
    except socket.error:
        pass
        
    return {"blocked": False, "matched": False}

def _block(url):
    return {
        "blocked": True,
        "matched": True,
        "type": "SSRF",
        "confidence": 0.99,
        "cwe": "CWE-918",
        "payload": url
    }

File: packages/agent/src/detection/ssrf.ts
import { TaintContext, isTainted } from '../taint/context';

export function detectSSRF(url: string, ctx?: TaintContext) {
  if (typeof url !== 'string') return { blocked: false, matched: false };
  const tainted = ctx ? isTainted(new String(url)) : false;
  if (!tainted) return { blocked: false, matched: false };

  try {
    const parsed = new URL(url);
    const host = parsed.hostname;
    const blockedHosts = ['localhost', '127.0.0.1', '169.254.169.254'];
    
    if (blockedHosts.includes(host) || host.endsWith('.internal')) {
      return {
        blocked: true, matched: true, type: 'SSRF', confidence: 0.99,
        cwe: 'CWE-918', payload: url, timestamp: Date.now()
      };
    }
  } catch (e) {}

  return { blocked: false, matched: false };
}